File: packages/solver/src/pairTable.ts
/**
 * Distance table for one F2L pair, ignoring everything else on the cube.
 *
 * A corner across eight slots in three orientations, with an edge across twelve slots in two, is
 * 576 positions — every one reachable, with a maximum distance of 6. Small enough to build on
 * demand without noticing.
 *
 * Ignoring the rest of the cube is what makes it a *lower bound* rather than an answer: bringing
 * the pair home while leaving the cross standing can only cost more than bringing it home with a
 * free hand. That is exactly what an admissible heuristic needs to be, and it is used both for
 * the pair being inserted and for every pair that has to survive the insertion — restoring a
 * disturbed pair cannot cost less than its distance either.
 */
import { applyMoves, CubeState } from "@cubing-companion/engine";
import { isSlotSolved, type Slot } from "@cubing-companion/analysis";
import { SEARCH_MOVES } from "./moves.ts";

/** Corner slot × orientation × edge slot × orientation. */
export const PAIR_INDEX_SPACE = 8 * 3 * 12 * 2;
export const MAX_PAIR_DISTANCE = 6;

const UNREACHABLE = 255;

/**
 * Where each move sends a corner or edge, and how it twists or flips it.
 *
 * Read out of the engine by applying each move to a solved cube, the same technique
 * `crossTable.ts` uses — `cp[i] === j` means slot `i` now holds what was at slot `j`, so `j`
 * travelled to `i`.
 */
const CORNER_TO: readonly Uint8Array[] = SEARCH_MOVES.map((move) => {
  const after = applyMoves(CubeState.solved(), [move]);
  const to = new Uint8Array(8);
  for (let i = 0; i < 8; i++) to[after.cp[i]!] = i;
  return to;
});

const CORNER_TWIST: readonly Uint8Array[] = SEARCH_MOVES.map((move) => {
  const after = applyMoves(CubeState.solved(), [move]);
  const twist = new Uint8Array(8);
  for (let i = 0; i < 8; i++) twist[after.cp[i]!] = after.co[i]!;
  return twist;
});

const EDGE_TO: readonly Uint8Array[] = SEARCH_MOVES.map((move) => {
  const after = applyMoves(CubeState.solved(), [move]);
  const to = new Uint8Array(12);
  for (let i = 0; i < 12; i++) to[after.ep[i]!] = i;
  return to;
});

const EDGE_FLIP: readonly Uint8Array[] = SEARCH_MOVES.map((move) => {
  const after = applyMoves(CubeState.solved(), [move]);
  const flip = new Uint8Array(12);
  for (let i = 0; i < 12; i++) flip[after.ep[i]!] = after.eo[i]!;
  return flip;
});

/** Pack a pair position into an index. */
export function packPair(
  cornerSlot: number,
  cornerOrientation: number,
  edgeSlot: number,
  edgeOrientation: number,
): number {
  return ((cornerSlot * 3 + cornerOrientation) * 12 + edgeSlot) * 2 + edgeOrientation;
}

/** Where a pair's two pieces currently sit, in an already-normalised state. */
export function pairIndexOf(state: CubeState, slot: Slot): number {
  let cornerSlot = -1;
  let edgeSlot = -1;
  for (let i = 0; i < 8; i++) {
    if (state.cp[i] === slot.corner) {
      cornerSlot = i;
      break;
    }
  }
  for (let i = 0; i < 12; i++) {
    if (state.ep[i] === slot.edge) {
      edgeSlot = i;
      break;
    }
  }
  return packPair(
    cornerSlot,
    state.co[cornerSlot]!,
    edgeSlot,
    state.eo[edgeSlot]!,
  );
}

/**
 * `pairIndexOf` for a caller that already knows where every piece is.
 *
 * `pairIndexOf` searches `cp` and `ep` for one piece, which is fine once but wasteful in a
 * search that asks about four pairs at every node — four scans of the same two arrays. Inverting
 * the permutations once per node and reading them costs less from the second pair onwards. The
 * inverses are the obvious ones: `inverseCp[p]` is the slot holding piece `p`.
 */
export function pairIndexFrom(
  state: CubeState,
  slot: Slot,
  inverseCp: Uint8Array,
  inverseEp: Uint8Array,
): number {
  const cornerSlot = inverseCp[slot.corner]!;
  const edgeSlot = inverseEp[slot.edge]!;
  return packPair(cornerSlot, state.co[cornerSlot]!, edgeSlot, state.eo[edgeSlot]!);
}

function build(slot: Slot): Uint8Array {
  const distance = new Uint8Array(PAIR_INDEX_SPACE).fill(UNREACHABLE);
  const solvedIndex = packPair(slot.corner, 0, slot.edge, 0);
  distance[solvedIndex] = 0;

  // Breadth-first over the four numbers directly, never touching a cube state.
  let frontier: [number, number, number, number][] = [[slot.corner, 0, slot.edge, 0]];
  let depth = 0;
  while (frontier.length > 0) {
    const next: [number, number, number, number][] = [];
    for (const [cornerSlot, cornerOrientation, edgeSlot, edgeOrientation] of frontier) {
      for (let m = 0; m < SEARCH_MOVES.length; m++) {
        const nextCorner = CORNER_TO[m]![cornerSlot]!;
        const nextTwist = (cornerOrientation + CORNER_TWIST[m]![cornerSlot]!) % 3;
        const nextEdge = EDGE_TO[m]![edgeSlot]!;
        const nextFlip = edgeOrientation ^ EDGE_FLIP[m]![edgeSlot]!;
        const index = packPair(nextCorner, nextTwist, nextEdge, nextFlip);
        if (distance[index] !== UNREACHABLE) continue;
        distance[index] = depth + 1;
        next.push([nextCorner, nextTwist, nextEdge, nextFlip]);
      }
    }
    frontier = next;
    depth++;
  }
  return distance;
}

const cache = new Map<string, Uint8Array>();

/** The table for one slot, built on first use and kept. */
export function pairTable(slot: Slot): Uint8Array {
  const key = `${slot.corner}:${slot.edge}`;
  let table = cache.get(key);
  if (!table) {
    table = build(slot);
    cache.set(key, table);
  }
  return table;
}

/** Lower bound on the moves needed to bring this pair home, from a normalised state. */
export function pairDistance(state: CubeState, slot: Slot): number {
  return pairTable(slot)[pairIndexOf(state, slot)]!;
}

// `isSlotSolved` is re-exported from `analysis` rather than redefined here: two copies of the
// same predicate are two things that can drift apart.
export { isSlotSolved };

File: swift/CubingCore/Sources/CubingCore/PairTable.swift
/// Exact lower bounds for one corner and edge pair, independent of the rest of the cube.
public enum PairTables {
    public static let indexSpace = 8 * 3 * 12 * 2
    public static let maxDistance = 6
    public static let unreachable: UInt8 = 255

    private static let cornerTo: [Int] = SearchMoves.all.flatMap { move in
        let after = CubeState.after([move])
        var to = [Int](repeating: 0, count: 8)
        for i in 0..<8 { to[Int(after.cp[i])] = i }
        return to
    }
    private static let cornerTwist: [Int] = SearchMoves.all.flatMap { move in
        let after = CubeState.after([move])
        var twist = [Int](repeating: 0, count: 8)
        for i in 0..<8 { twist[Int(after.cp[i])] = Int(after.co[i]) }
        return twist
    }

    public static func pack(_ corner: Int, _ twist: Int, _ edge: Int, _ flip: Int) -> Int {
        ((corner * 3 + twist) * 12 + edge) * 2 + flip
    }

    @inline(__always)
    static func step(_ index: Int, _ move: Int) -> Int {
        let flip = index & 1
        let edge = (index >> 1) % 12
        let twist = (index / 24) % 3
        let corner = index / 72
        let ci = move * 8 + corner
        let ei = move * 12 + edge
        return pack(
            cornerTo[ci], (twist + cornerTwist[ci]) % 3,
            CrossTables.edgeTo[ei], flip ^ CrossTables.edgeFlip[ei])
    }

    static func index(_ state: CubeState, _ slot: Slot) -> Int {
        let corner = (0..<8).first { Int(state.cp[$0]) == slot.corner }!
        let edge = (0..<12).first { Int(state.ep[$0]) == slot.edge }!
        return pack(corner, Int(state.co[corner]), edge, Int(state.eo[edge]))
    }

    static func index(_ state: CubeState, _ slot: Slot, _ inverseCp: [Int], _ inverseEp: [Int]) -> Int {
        let corner = inverseCp[slot.corner], edge = inverseEp[slot.edge]
        return pack(corner, Int(state.co[corner]), edge, Int(state.eo[edge]))
    }

    private static func build(_ slot: Slot) -> [UInt8] {
        var distance = [UInt8](repeating: unreachable, count: indexSpace)
        let solved = pack(slot.corner, 0, slot.edge, 0)
        distance[solved] = 0
        var frontier = [solved]
        var depth: UInt8 = 0
        while !frontier.isEmpty {
            var next: [Int] = []
            for index in frontier {
                for move in 0..<18 {
                    let child = step(index, move)
                    if distance[child] != unreachable { continue }
                    distance[child] = depth + 1
                    next.append(child)
                }
            }
            frontier = next
            depth += 1
        }
        return distance
    }

    // Twenty-four small tables, built once. Immutable after initialization, so readers need no lock.
    private static let tables: [Int: [UInt8]] = {
        var result: [Int: [UInt8]] = [:]
        for slot in geometry.flatMap(\.slots) { result[slot.corner * 12 + slot.edge] = build(slot) }
        return result
    }()

    public static func table(_ slot: Slot) -> [UInt8] { tables[slot.corner * 12 + slot.edge]! }

    /// The state may be held in any orientation.
    public static func distance(_ state: CubeState, _ slot: Slot) -> Int {
        let normalised = state.normalized
        return Int(table(slot)[index(normalised, slot)])
    }
}

public func pairDistance(_ state: CubeState, _ slot: Slot) -> Int {
    PairTables.distance(state, slot)
}
